### PR TITLE
Bug 1637302: Compress artifact uploads

### PR DESCRIPTION
**Pull request by catlee**
Bugzilla Bug: [1637302](https://bugzilla.mozilla.org/show_bug.cgi?id=1637302)

### CODE REVIEW

**Comment by imbstack** [comment]
@catlee If you rebase on master and push, the tests are working for docker-worker now.

Also I'll set you up with scopes to run the tests locally. I'll ping you when it is all set up.

File: workers/docker-worker/src/lib/upload_to_s3.js [hunk 1]
```diff
@@ -40,12 +42,21 @@ module.exports = async function uploadToS3 (
       await tmp.writeFile(source);
     } else {
       let stream = fs.createWriteStream(tmp.path);
-      await pipe(source, stream);
+      // TODO: Do we care if this is larger than the original?
```
Comment (catlee) [inline workers/docker-worker/src/lib/upload_to_s3.js:45]: I suspect we don't care. I noticed this in tests where the original file is very small (5 bytes) and the resulting compressed file is 25 bytes.

File: workers/docker-worker/src/lib/upload_to_s3.js [hunk 2]
```diff
@@ -40,12 +42,21 @@ module.exports = async function uploadToS3 (
...
+        debug(`compressed ${gzip.bytesWritten} bytes`);
+      } else {
+        await pipe(source, stream);
+      }
     }
-    debug(`wrote source file to ${tmp.path} for ${artifactName}`);
-
     let stat = await fs.stat(tmp.path);
     size = stat.size;
+    httpsHeaders['content-length'] = size;
```
Comment (catlee) [inline workers/docker-worker/src/lib/upload_to_s3.js:57]: Is this safe to do? Can we stop setting content-length in the callers?

File: workers/docker-worker/src/lib/features/artifacts.js [hunk 3]
```diff
@@ -143,7 +154,7 @@ class Artifacts {
         let start = process.hrtime();
 
         let {digest, size} = await uploadToS3(taskHandler.queue, taskId, runId, stream,
-          entryName, expiry, headers);
+          entryName, expiry, headers, null, null, compress);
```
Comment (catlee) [inline workers/docker-worker/src/lib/features/artifacts.js:157]: I wasn't sure if there was a more idiomatic way of passing in the compress flag without these extra null parameters.

**Comment by catlee** [comment]
TODO: write some tests for this behaviour

File: workers/docker-worker/src/lib/upload_to_s3.js (hunk 2 above)
Comment (djmitche) [reply workers/docker-worker/src/lib/upload_to_s3.js:57]: Yes, it's safe.

File: workers/docker-worker/test/integration/file_artifact_test.js [hunk 4]
```diff
@@ -337,4 +339,52 @@ suite('artifact extraction tests', () => {
 
     assert.ok(retry);
   });
+
+  test('automatic gzip compression', async () => {
```
[the remark this answers is not in the document]
Comment (djmitche) [reply workers/docker-worker/test/integration/file_artifact_test.js:350]: This covers both the compresed and not-compressed cases.

**Comment by djmitche** [comment]
> This covers both the compresed and not-compressed cases.

(I got distracted and didn't finish that comment)

.. so I think these tests are sufficient.

**Pull request merged**